include/swift/ABI/Metadata.h

```diff
@@ -1170,6 +1170,26 @@ struct TargetClassMetadata : public TargetAnyClassMetadata<Runtime> {
 
   /// Is this class an artificial subclass, such as one dynamically
   /// created for various dynamic purposes like KVO?
+  //
+  // [NOTE: Dynamic-subclass-KVO]
+  // To implement Key-Value Observing without any code that notifies the
+  // observer, the KVO infrastructure uses dynamic subclassing with Objective-C
+  // runtime. When a variable is observed, KVO creates a secret dynamic subclass
+  // of that class under the hood which are defined with a prefix of
+  // `NSKVONotifying_`.
+  //
+  // While the observed variables have the type of the dynamic subclass, they
+  // must appear like their non-observed counterparts for the front-end user. To
+  // achieve this, the dynamic subclass overrides `-class` method which returns
+  // the original class type, and internally refers to the subclass in the
+  // runtime.
+  //
+  // In the created subclass, `-set` methods for observed variables are
+  // overridden, where the calls to the observer notifications are triggered.
+  // KVO only generates one dynamic subclass for each class which overrides all
+  // setter methods of variables being observed. That is, setters of variables
+  // that are not observed are also not overridden in the dynamic subclass for
+  // efficiency.
```

Comment:
While, I think this is generally useful, I think it would be nice if we could shorten it, since some aspects of the text may already be clear to someone working on the runtime. We can link article for a longer explanation.

How about something like:

> KVO allows modifying object behavior without modifying the object's code.
> This is done by dynamically creating a subclass of the the object's type.
> The isa pointer on the observed object is swapped out to point to the artificial
> subclass, which has the following properties:
> - Setters for observed keys are overriden to additionally post notifications.
> - The `-class` method is overriden to return the original class type instead of the
>    artificial subclass type.
>
> For more details, see: https://www.mikeash.com/pyblog/friday-qa-2009-01-23.html

Also, it's probably better to move the note above the doc comment and add a `See [NOTE: Dynamic-subclass-KVO]` here too. Otherwise, when you try to see the docs, Xcode will not associate the `isArtificialSubclass` method with the doc comment since there is a non-doc comment in between the doc comment and the method signature.

Reply:
cc @mikeash in case you have additional thoughts on wording/context